File: src/main/java/org/team401/robot/MotionProfiles.java
package org.team401.robot;

import java.io.File;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Scanner;

import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * Static class that reads motion profiles from .csv files.
 * Data expected is 3 columns:
 * Position(rotations), Velocity(RPM), Duration(ms)
 */

public class MotionProfiles {
	//alliance param defaults to false
	public static double[][][] get(String start, String tgt, boolean mecanum){
		return get(start, tgt, mecanum, false);
	}

	/**
	 * Scans multiple .csv files to get the correct motion profiles for any given start and end point.
	 *
	 * @param start Where the robot starts at the beginning of the match.  Valid inputs are "L", "C", and "R".
	 * @param tgt Where the robot will drive.  Valid inputs are "LH", "LL", "CL", "RL", and "RH".
	 * @param mecanum Whether to read spreadsheets for each gearbox or for left and right sets of gearboxes.
	 * @param alliance Whether or not the alliance we're on matters.
	 * @return Array with length 4 or 2 containing complete motion profiles.
	 */
	public static double[][][] get(String start, String tgt, boolean mecanum, boolean alliance) {
		//Adds an alliance identifier("R", "B", or "I") if the path is alliance-dependent. "I" should never occur.
		String all = alliance||tgt.endsWith("H") ? DriverStation.getInstance().getAlliance().name().substring(0, 1) : "";

		//Constructs the .csv names from parameters
		return mecanum ?
			new double[][][]{
				//Mecanum drive
				scanCSV("/home/mecanumprofiles/"+start+tgt+all+" FL"),
				scanCSV("/home/mecanumprofiles/"+start+tgt+all+" FR"),
				scanCSV("/home/mecanumprofiles/"+start+tgt+all+" RL"),
				scanCSV("/home/mecanumprofiles/"+start+tgt+all+" RR")
			}: new double[][][]{
				//Tank drive
				scanCSV("/home/tankprofiles/"+start+tgt+all+" L"),
				scanCSV("/home/tankprofiles/"+start+tgt+all+" R")
			};
	}

	/**
	 * Parses the contents of a .csv as a 2D array of doubles.
	 * @param fileName Name, without file extension, of the file to read from.
	 * @return Contents of the .csv
	 */
	private static double[][] scanCSV(String fileName){
		//Add file extension
		fileName = fileName + ".csv";

		//BufferedReader will read from the file, Scanner will find row count
		BufferedReader br;
		Scanner scan;

		//Index of the loops
		int i = 0;

		//Places to store each line and the finished product
		String[] str;
		double[][] result;

		//File I/O sometimes causes errors
		try {
			//Scan the file into Java
			br = new BufferedReader(new FileReader(fileName));

			//Size the result to the length of the file
			scan = new Scanner(new File(fileName));
			scan.useDelimiter("\n");
			while(scan.hasNext()){
				scan.next();
				i++;
			}
			result = new double[i][3];

			//Read each line and parse it into 3 doubles
			i = 0;
			for(String line; (line=br.readLine())!=null; i++) {
				str = line.split(",");
				for (int j = 0; j < 3; j++)
					result[i][j] = Double.parseDouble(str[j]);
			}

			//Close the file reader and return the result
			br.close();
			return result;

		//Notify user of exceptions
		} catch (FileNotFoundException e) {
			notifyError("File " + fileName+" not found!");
		} catch (IOException e) {
			notifyError("IOException in scanCSV while scanning " + fileName + "!");
		}

		//Return empty array if there was an exception
		return new double[0][0];
	}
	private static void notifyError(String message){
		SmartDashboard.putString("Latest Error", message);
		System.out.println(message);
	}
}

File: src/main/java/org/team401/robot/auto/modes/MotionProfiles.kt
package org.team401.robot.auto.modes

import com.ctre.CANTalon
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import org.team401.robot.auto.AutoModeSelector
import java.io.*
import java.util.*



internal class MotionProfiles(profile: String){
//come back to this
private var setValue: CANTalon.SetValueMotionProfile = CANTalon.SetValueMotionProfile.Disable
private val status = CANTalon.MotionProfileStatus(); // tracks the status of our MP Update with every iteration of the loop
private val talon = CANTalon(1) // the talons running the MP
private var state: Int = 0 // state for the state machine
private var loopTimeOut: Int = 0
private val kLoopTimeOut = 10
private val minPoints: Int = 10
private val path = scanCSV(profile)
    
    /*
    Plan:
    Set this us so routine() runs a motion profile

    runs every iteration of a auto periodic loop

    if motionprofiles on?
        No
            state = 0
            looptimeout = -1 //disable
        Yes
            //run a MP
            when(state){
                0 ->
                    startFilling() //buffer the points into the talons
                    setvalue = disabled
                    looptimeout = kloopstimeout //tracks progress
                    state = 1
                1 ->
                    //check if enough points have been buffered'
                    if (statud.btmbuffercount > minpoints){
                        setValue = enabled //starts profile once
                        state = 2;
                    }
                2 ->
                    //check how the profile is going
                    if(status.isunderrun == false){
                        looptimeout = klooptimeout
                    }
                    if(status.activepointValid && status.activepointislast)//if it is on the last point/ done
                        setvalue = hold
                        state = 0//reset state machine
                        looptimeout = -1 //disable
            }
     */

    private fun control(){
        if (talon.controlMode != CANTalon.TalonControlMode.MotionProfile) {
            state = 0
            loopTimeOut = -1
        }else {
            when (state) {
                0 -> {
                    startFilling(path)
                    setValue = CANTalon.SetValueMotionProfile.Disable
                    state = 1
                    loopTimeOut = kLoopTimeOut
                }
                1->{
                    if(status.btmBufferCnt >= minPoints)
                        setValue = CANTalon.SetValueMotionProfile.Enable
                        state = 2
                }
                2->{
                    if(status.hasUnderrun == false)
                        loopTimeOut = kLoopTimeOut
                    if(status.activePointValid && status.activePoint.isLastPoint)
                        setValue = CANTalon.SetValueMotionProfile.Hold
                        state = 0
                        loopTimeOut = -1
                }
            }
        }
    }


}

/**
 * Parses the contents of a .csv as a 2D array of doubles.
 * @param fileName Name, without file extension, of the file to read from.
 * @return Contents of the .csv
 */
private fun scanCSV(fileName: String): Array<DoubleArray> {
    var fileName = fileName
    //Add file extension
    fileName = fileName + ".csv"

    //BufferedReader will read from the file, Scanner will find row count
    val br: BufferedReader
    val scan: Scanner

    //Index of the loops
    var i = 0

    //Places to store each line and the finished product
    var str: Array<String>
    val result: Array<DoubleArray>

    //File I/O sometimes causes errors
    try {
        //Scan the file into Java
        br = BufferedReader(FileReader(fileName))

        //Size the result to the length of the file
        scan = Scanner(File(fileName))
        scan.useDelimiter("\n")
        while (scan.hasNext()) {
            scan.next()
            i++
        }
        result = Array(i) { DoubleArray(3) }

        //Read each line and parse it into 3 doubles
        i = 0
        var line: String = br.readLine()
        while (line != null) {
            str = line.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (j in 0..2)
                result[i][j] = java.lang.Double.parseDouble(str[j])
            i++
        }

        //Close the file reader and return the result
        br.close()
        return result

        //Notify user of exceptions
    } catch (e: FileNotFoundException) {
        notifyError("File $fileName not found!")
    } catch (e: IOException) {
        notifyError("IOException in scanCSV while scanning $fileName!")
    }

    //Return empty array if there was an exception
    return Array(0) { DoubleArray(0) }
}

private fun notifyError(message: String) {
    SmartDashboard.putString("Latest Error", message)
    println(message)
}

//work in progress
private fun startFilling(profile: Array<DoubleArray>, totalRows: Int = profile.size){

    val point: CANTalon.TrajectoryPoint




}
